onmt/modules/RecurrentBatchNorm.lua

```diff
@@ -0,0 +1,91 @@
+require('nngraph')
+
+--[[
+An implementation of recurrent batch normalization as explained in:
+https://arxiv.org/abs/1603.09025
+
+We compute per-time-step batch statistics as part of normalization. Each time step's
+corresponding nn.BatchNormalization module is stored in self.modules, and the table
+self.modules is *shared across all clones*.
+--]]
+
+local RecurrentBatchNorm, parent = torch.class('onmt.RecurrentBatchNorm', 'nn.Container')
+
+--[[
+Parameters:
+
+  * `dim` - The dimension of the input
+  * `eps` - Small number to add to avoid divide-by-zero (defaults to 1e-5)
+  * `momentum` - Batch normalization momentum (defaults to 0.1)
+  * `affine` - Whether to learn a scaling and bias term (see nn.BatchNormalization for details)
+  * `noBias` - Whether to hold the bias to 0 (but learn the scaling term)
+--]]
+function RecurrentBatchNorm:__init(dim, eps, momentum, affine, noBias)
+  parent.__init(self)
+
+  -- store variables for later reconstruction
+  self.dim = dim
+  self.eps = eps
+  self.momentum = momentum
+  self.affine = affine
+  self.noBias = noBias
+
+  self:setTimeStep(1)
+end
+
+function RecurrentBatchNorm:__timeStepClone()
+  local bn = nn.BatchNormalization(self.dim, self.eps, self.momentum, self.affine)
+  if self._type == 'torch.CudaTensor' then
```

Comment:
It is more generic to do:

```lua
bn:type(self._type)
```

And it will also fix CPU execution as we are working with `float`.